### ISSUE TITLE
Question on Memory Addressing Scheme

### ISSUE THREAD

**Issue by itzmeanjan** [maticnetwork/miden#116]
Hi,

I was reading Miden Assembly specification's RAM addressing [modes](https://hackmd.io/YDbjUVHTRn64F4LPelC-NA#Random-access-memory)

In context of function local memory and absolute index based memory addressing, I was wondering how do I handle following depicted scenario.

```bash
# Imagine I've following Miden Assembly code, where I'd like to use local memory ( reserved while
# invoking `fn0` ) in `fn1` procedure, which is invoked from `fn0`
#
# i ) But following program instead accesses memory locations by absolute indices [ not desired ]
# ii ) If I declare `fn2`, instead of `fn1`, then it reserves another 2 VM words; lets me work with that [ not desired ]
#
# Question: How do I refer `fn{1,2}` to use local memory reserved while invoking their parent function `fn0` ?

    proc.fn2.2
        pushw.local.0
        pushw.local.1
    end

    proc.fn1.0
        pushw.mem.0
        pushw.mem.1
    end

    proc.fn0.2
        push.3.2.1.0
        push.7.6.5.4

        popw.local.1
        popw.local.0

        exec.fn1
    end

    begin
        exec.fn0
    end
```

**Comment by itzmeanjan** [comment]
I just noticed another thing, when following Miden Assembly is executed

```bash
    proc.fn1.0
        pushw.mem.1
        pushw.mem.2

        dup.1
        u32add.unsafe
        drop

        popw.mem.2
        popw.mem.1
    end

    proc.fn0.2
        push.3.2.1.0
        push.7.6.5.4

        popw.local.1
        popw.local.0

        exec.fn1
       
       # imagine here fn0 uses local.{0,1} for some reason ! #
    end

    begin
        exec.fn0

        pushw.mem.1
        pushw.mem.2
    end
```

It terminates with stack trace

```bash
[1, 1, 2, 3, 4, 5, 6, 7, 0, 0, 0, 0, 0, 0, 0, 0]
```

So it seems to be pretty easy to damage `fn0`'s reserved memory allocation, from `fn1` --- a function which said it won't use any local memory but it can damage its caller's working memory area.

Am I misunderstanding something ?

**Comment by itzmeanjan** [comment]
> The above is relevant only for "trusted procedures". In the future, we'll also have "untrusted procedures" (invoked via call instruction), and these will have disjointed memories.

Yes, this distinction between `{trusted|untrusted}` procedures makes sense.

**Comment by bobbinth** [comment]
Closing this as support for retrieving absolute memory addresses for locals was implemented in #124
